src/main/bootstrap/cache_validator.rs

```diff
@@ -0,0 +1,198 @@
+use crate::main::{CrustData, EventLoopCore};
+use crate::nat::GetExtAddr;
+use mio::{Poll, PollOpt, Ready, Token};
+use mio_extras::channel::{channel, Receiver, Sender};
+use safe_crypto::{PublicEncryptKey, SecretEncryptKey};
+use std::any::Any;
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::net::SocketAddr;
+use std::rc::{Rc, Weak};
+
+/// When peer is tested, if it's still online, we send an encrypted request and wait for response.
+/// This is the maximum time we wait before declaring that peer is dead.
+const PEER_TEST_TIMEOUT_SEC: u64 = 30;
+
+/// Bootstrap cache validator tests inactive cached peers if they are still alive.
+/// It receives peers from async in-memory channel.
+/// Cache validator is a future/state that never finishes.
+pub struct CacheValidator {
+    token: Token,
+    our_pk: PublicEncryptKey,
+    our_sk: SecretEncryptKey,
+    expired_rx: Receiver<HashSet<PeerInfo>>,
+    expired_tx: Sender<HashSet<PeerInfo>>,
+    /// Tokens for in progress states testing peer connectivity.
+    sent_requests: HashSet<Token>,
+    /// When need to reference self from the callbacks.
+    self_weak: Weak<RefCell<CacheValidator>>,
+}
+
+impl CacheValidator {
+    /// Starts running bootstrap cache entry validation.
+    pub fn start(
+        core: &mut EventLoopCore,
+        poll: &Poll,
+        our_pk: PublicEncryptKey,
+        our_sk: SecretEncryptKey,
+    ) -> crate::Res<()> {
+        let token = core.get_new_token();
+
+        let (expired_tx, expired_rx) = channel();
+        core.user_data_mut().expired_cached_peers_tx = Some(expired_tx.clone());
+        poll.register(&expired_rx, token, Ready::readable(), PollOpt::edge())?;
+
+        let state = Rc::new(RefCell::new(Self {
+            token,
+            our_pk,
+            our_sk,
+            expired_rx,
+            expired_tx: expired_tx.clone(),
+            sent_requests: Default::default(),
+            self_weak: Default::default(),
+        }));
+        state.borrow_mut().self_weak = Rc::downgrade(&state);
+        let _ = core.insert_state(token, state.clone());
+
+        Ok(())
+    }
+
+    /// Sends STUN requests to cached peers that recently been inactive.
+    fn ping_inactive_peers(
+        &mut self,
+        core: &mut EventLoopCore,
+        poll: &Poll,
+        inactive_peers: HashSet<PeerInfo>,
+    ) {
+        debug!("Testing {} inactive cached peers.", inactive_peers.len());
+        for peer in inactive_peers {
+            self.ping_peer_and_cache_alive(core, poll, peer);
+        }
+    }
+
+    /// Send STUN request to given peer and wait for response.
+    /// If response arrives within timeout, peer is added to the bootstrap cache head.
+    fn ping_peer_and_cache_alive(&mut self, core: &mut EventLoopCore, poll: &Poll, peer: PeerInfo) {
+        let self_weak = self.self_weak.clone();
+        let expired_tx = self.expired_tx.clone();
+        let on_done = move |core: &mut EventLoopCore,
```

Comment:
These were all called `finish` (and it wasn't me who did that, it probably belonged to 2016 naming convention set by Andrew). Why change all that ? Pls revert back to `finish` and report discrepancies if you find any as that should also be corrected. I don't remember `on_done` anywhere in Crust before so pls refrain from changing those.

Reply:
`finish` is an imperative form and `on_finish`/`on_done` sounds better to me as that implies callback. Anyway this is a minor thing and I'll update it. But having something settled is not the best argument to avoid refactoring to me :)

Reply:
> But having something settled is not the best argument to avoid refactoring to me :)

Sounds good for an oxymoron :P